apiserver/client/status.go

```diff
@@ -574,67 +580,54 @@ func (m machineAndContainers) Containers(id string) []*state.Machine {
 	return m[id][1:]
 }
 
-func processMachines(
-	idToMachines map[string][]*state.Machine,
-	idToIpAddresses map[string][]*state.Address,
-	idToDeviceToSpaces map[string]map[string]set.Strings,
-	idToLinkLayerDevices map[string][]*state.LinkLayerDevice,
-) map[string]params.MachineStatus {
+func (c *statusContext) processMachines() map[string]params.MachineStatus {
 	machinesMap := make(map[string]params.MachineStatus)
 	cache := make(map[string]params.MachineStatus)
-	for id, machines := range idToMachines {
+	for id, machines := range c.machines {
 
 		if len(machines) <= 0 {
 			continue
 		}
 
 		// Element 0 is assumed to be the top-level machine.
 		tlMachine := machines[0]
-		hostStatus := makeMachineStatus(
-			tlMachine,
-			idToIpAddresses[tlMachine.Id()],
-			idToDeviceToSpaces[tlMachine.Id()],
-			idToLinkLayerDevices[tlMachine.Id()],
-		)
+		hostStatus := c.makeMachineStatus(tlMachine)
 		machinesMap[id] = hostStatus
 		cache[id] = hostStatus
 
 		for _, machine := range machines[1:] {
 			parent, ok := cache[state.ParentId(machine.Id())]
 			if !ok {
-				panic("We've broken an assumpution.")
+				logger.Errorf("programmer error, please file a bug, reference this whole log line: %q, %q", id, machine.Id())
+				continue
 			}
 
-			status := makeMachineStatus(
-				machine,
-				idToIpAddresses[machine.Id()],
-				idToDeviceToSpaces[machine.Id()],
-				idToLinkLayerDevices[machine.Id()],
-			)
+			status := c.makeMachineStatus(machine)
 			parent.Containers[machine.Id()] = status
 			cache[machine.Id()] = status
 		}
 	}
 	return machinesMap
 }
 
-func makeMachineStatus(
-	machine *state.Machine,
-	ipAddresses []*state.Address,
-	spaces map[string]set.Strings,
-	linkLayerDevices []*state.LinkLayerDevice,
-) (status params.MachineStatus) {
+func (c *statusContext) makeMachineStatus(machine *state.Machine) (status params.MachineStatus) {
+	machineID := machine.Id()
+	ipAddresses := c.ipAddresses[machineID]
+	spaces := c.spaces[machineID]
+	linkLayerDevices := c.linkLayerDevices[machineID]
+
```

Comment:
A lot of this data never gets rendered in tabular format.
I wonder if we should move away from a FullStatus that generates everything and a renderer that filters out some of it.